File: app/ios/Runner/WatchBridge.swift
/// Phone side of the Apple Watch bridge.
///
/// Owns the `WCSession`, owns the two Flutter channels, and buffers across the gap
/// between them. The gap is the whole reason this class exists: WatchConnectivity can
/// launch this app into the background purely to deliver a watch message, and at that
/// moment the Flutter engine may not exist yet, let alone have run `main()`. Anything
/// arriving before Dart is ready is parked in `pendingToDart` and replayed when it is.
///
/// Wire format is documented in README.md ("Apple Watch Companion"). Dart counterpart:
/// `app/lib/watch_bridge.dart`. Watch counterpart: `app/ios/WatchApp/Sources/WatchSession.swift`.
import Flutter
import Foundation
import Speech
import WatchConnectivity

final class WatchBridge: NSObject {
  static let shared = WatchBridge()

  private static let methodChannelName = "org.marsaprs/watch"
  private static let eventChannelName = "org.marsaprs/watch/events"

  /// How long a watch `sendMessage` reply handler may be held open while we wait for
  /// Dart to perform the send. WatchConnectivity's own reply window is short and
  /// unforgiving, so we answer "queued" well before it expires rather than risk the
  /// watch seeing a transport error for a send that is actually in flight.
  private static let replyWatchdog: TimeInterval = 8

  private var methodChannel: FlutterMethodChannel?
  private var eventSink: FlutterEventSink?

  /// Events that arrived before Dart called `ready`. Ordered; replayed verbatim.
  private var pendingToDart: [[String: Any]] = []
  /// clientId → the watch's reply handler, until Dart reports the result.
  private var pendingReplies: [String: ([String: Any]) -> Void] = [:]
  /// Last context we pushed, re-sent when the watch becomes reachable again.
  private var lastContext: [String: Any]?

  private var session: WCSession? { WCSession.isSupported() ? WCSession.default : nil }

  // ── lifecycle ───────────────────────────────────────────────────────────────

  /// Must be called from `didFinishLaunchingWithOptions` *before* anything slow.
  /// If no delegate is assigned by the time launch returns, a message that woke the
  /// app is dropped and the watch's reply handler times out.
  func activate() {
    guard WCSession.isSupported() else { return } // false on iPad
    let s = WCSession.default
    s.delegate = self
    s.activate()
  }

  /// Called once the implicit Flutter engine exists. Safe to call again after an
  /// engine restart; the channels are simply rebuilt.
  func attach(messenger: FlutterBinaryMessenger) {
    let method = FlutterMethodChannel(name: Self.methodChannelName, binaryMessenger: messenger)
    method.setMethodCallHandler { [weak self] call, result in
      self?.handle(call: call, result: result)
    }
    methodChannel = method

    FlutterEventChannel(name: Self.eventChannelName, binaryMessenger: messenger)
      .setStreamHandler(self)
  }

  // ── Dart → here ─────────────────────────────────────────────────────────────

  private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "ready":
      flushPendingToDart()
      result(state())
    case "status":
      result(state())
    case "setContext":
      lastContext = args
      result(["ok": pushContext(args)])
    case "pushMessage":
      result(["transport": push(message: args)])
    case "pushMessages":
      let msgs = args["messages"] as? [[String: Any]] ?? []
      // One envelope, not N: transferUserInfo is FIFO but each call is a separate
      // wake of the watch app, and a catch-up burst would wake it many times over.
      result(["transport": msgs.isEmpty ? "none" : push(message: ["batch": msgs])])
    case "sendResult":
      deliver(sendResult: args)
      result(["ok": true])
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func state() -> [String: Any] {
    guard let s = session else {
      return ["supported": false, "paired": false, "appInstalled": false,
              "reachable": false, "activated": false]
    }
    return ["supported": true,
            "paired": s.isPaired,
            "appInstalled": s.isWatchAppInstalled,
            "reachable": s.isReachable,
            "activated": s.activationState == .activated]
  }

  // ── here → watch ────────────────────────────────────────────────────────────

  /// Latest-value-wins state: token, destination, conversation list, speak flag.
  /// Survives the watch app not running, and coalesces if we push faster than the
  /// link drains. Returns false when there is nothing to talk to.
  @discardableResult
  private func pushContext(_ dict: [String: Any]) -> Bool {
    guard let s = session, s.activationState == .activated, s.isPaired else {
      NSLog("[watch] context skipped (activated=\(session?.activationState == .activated) paired=\(session?.isPaired ?? false))")
      return false
    }
    do {
      try s.updateApplicationContext(sanitized(dict))
      NSLog("[watch] context pushed (sharing=\(dict["sharing"] ?? "?") lastId=\(dict["lastId"] ?? "?"))")
      return true
    } catch {
      NSLog("[watch] updateApplicationContext failed: \(error.localizedDescription)")
      return false
    }
  }

  /// A message (or a `batch` of them). `sendMessage` is the low-latency path when the
  /// watch app is up; `transferUserInfo` is the durable one — FIFO, persisted across
  /// termination on both sides, and it wakes the watch app in the background to
  /// receive. The watch dedupes by message id, so the fallback racing the fast path
  /// is harmless.
  private func push(message: [String: Any]) -> String {
    guard let s = session, s.activationState == .activated,
          s.isPaired, s.isWatchAppInstalled else { return "none" }
    let payload = sanitized(message)
    let id = (payload["id"] as? Int).map(String.init) ?? "batch"
    if s.isReachable {
      NSLog("[watch] push id=\(id) via sendMessage")
      s.sendMessage(payload, replyHandler: nil) { error in
        NSLog("[watch] sendMessage id=\(id) failed, queueing: \(error.localizedDescription)")
        s.transferUserInfo(payload)
      }
      return "sendMessage"
    }
    // Not reachable means the watch app is not frontmost. The transfer is durable
    // and will be delivered when it next runs, but it will arrive silently — this
    // log line is what distinguishes "the relay never fired" from "the watch was
    // not listening", which look identical from the wrist.
    NSLog("[watch] push id=\(id) via transferUserInfo (watch app not frontmost)")
    s.transferUserInfo(payload)
    return "userInfo"
  }

  /// WatchConnectivity only accepts property-list types, and `NSNull` is not one —
  /// a single null anywhere makes the whole call fail, taking every other key with
  /// it. Flutter encodes a Dart `null` as `NSNull`, and the fields most likely to be
  /// null (`token`, `destination`, a sender's short id) are null exactly when the
  /// user has not started sharing or opened a thread yet, so the failure lands on a
  /// first run and looks like the watch is simply dead. Absent means null by
  /// contract on the watch side, so dropping them loses nothing.
  private func sanitized(_ dict: [String: Any]) -> [String: Any] {
    var out: [String: Any] = [:]
    for (key, value) in dict {
      if value is NSNull { continue }
      if let nested = value as? [String: Any] {
        out[key] = sanitized(nested)
      } else if let list = value as? [[String: Any]] {
        out[key] = list.map(sanitized)
      } else {
        out[key] = value
      }
    }
    return out
  }

  /// The out-of-band answer to a watch send we could not answer inline.
  private func deliver(sendResult: [String: Any]) {
    var payload = sendResult
    payload["type"] = "sendResult"
    guard let clientId = sendResult["clientId"] as? String else {
      _ = push(message: payload)
      return
    }
    if let reply = pendingReplies.removeValue(forKey: clientId) {
      reply(payload) // still inside the watch's reply window
    } else {
      _ = push(message: payload) // watchdog already answered "queued"
    }
  }

  // ── watch → Dart ────────────────────────────────────────────────────────────

  private func emit(_ event: [String: Any]) {
    if let sink = eventSink {
      sink(event)
    } else {
      // Cap the buffer: a watch that keeps talking to an app that never finishes
      // launching must not grow this without bound.
      if pendingToDart.count >= 64 { pendingToDart.removeFirst() }
      pendingToDart.append(event)
    }
  }

  private func flushPendingToDart() {
    guard let sink = eventSink, !pendingToDart.isEmpty else { return }
    let queued = pendingToDart
    pendingToDart.removeAll()
    for e in queued { sink(e) }
  }

  /// Route an inbound watch payload to Dart, holding `reply` open until Dart answers
  /// or the watchdog fires — whichever comes first.
  private func route(_ payload: [String: Any], reply: (([String: Any]) -> Void)?) {
    // A push-to-talk clip sent inline rather than as a file transfer. Handled here
    // and never forwarded to Dart: turning audio into text is this side's job and
    // needs no Flutter engine, which matters because this can arrive while the app
    // is still starting in the background.
    if payload["type"] as? String == "talkAudio",
       let audio = payload["audio"] as? Data,
       let id = payload["clientId"] as? String {
      reply?(["received": true])
      receiveInlineAudio(audio, clientId: id)
      return
    }

    var event = payload
    let clientId = payload["clientId"] as? String

    if let reply, let clientId {
      pendingReplies[clientId] = reply
      DispatchQueue.main.asyncAfter(deadline: .now() + Self.replyWatchdog) { [weak self] in
        guard let self, let pending = self.pendingReplies.removeValue(forKey: clientId) else { return }
        // Not a failure. Dart is still working; the real answer arrives out of band.
        pending(["type": "sendResult", "clientId": clientId, "queued": true])
      }
    } else if let reply {
      reply(["ok": true])
    }

    if event["type"] == nil { event["type"] = "unknown" }
    emit(event)
  }

  private func emitWatchState() {
    var e = state()
    e["type"] = "watchState"
    emit(e)
  }
}

// ── WCSessionDelegate ─────────────────────────────────────────────────────────

extension WatchBridge: WCSessionDelegate {
  func session(_ session: WCSession, activationDidCompleteWith activationState: WCSessionActivationState,
               error: Error?) {
    if let error { NSLog("[watch] activation failed: \(error.localizedDescription)") }
    DispatchQueue.main.async {
      self.emitWatchState()
      if activationState == .activated, let ctx = self.lastContext { self.pushContext(ctx) }
    }
  }

  func sessionDidBecomeInactive(_ session: WCSession) {}

  /// Required on iOS: the user switched to a different paired watch. Without
  /// re-activating, this app is deaf to the new one for the rest of its life.
  func sessionDidDeactivate(_ session: WCSession) {
    WCSession.default.activate()
  }

  func sessionWatchStateDidChange(_ session: WCSession) {
    DispatchQueue.main.async { self.emitWatchState() }
  }

  func sessionReachabilityDidChange(_ session: WCSession) {
    DispatchQueue.main.async {
      self.emitWatchState()
      // The watch stops its own direct polling the moment we are reachable, so it
      // needs current state immediately rather than at the next natural push.
      if session.isReachable, let ctx = self.lastContext { self.pushContext(ctx) }
    }
  }

  func session(_ session: WCSession, didReceiveMessage message: [String: Any]) {
    DispatchQueue.main.async { self.route(message, reply: nil) }
  }

  func session(_ session: WCSession, didReceiveMessage message: [String: Any],
               replyHandler: @escaping ([String: Any]) -> Void) {
    DispatchQueue.main.async { self.route(message, reply: replyHandler) }
  }

  func session(_ session: WCSession, didReceiveUserInfo userInfo: [String: Any] = [:]) {
    DispatchQueue.main.async { self.route(userInfo, reply: nil) }
  }

  func session(_ session: WCSession, didReceiveApplicationContext applicationContext: [String: Any]) {
    DispatchQueue.main.async { self.route(applicationContext, reply: nil) }
  }

  /// A push-to-talk clip recorded on the watch.
  ///
  /// The watch has no speech recogniser — `Speech.framework` does not exist on
  /// watchOS — so it records and this side listens. The file must be copied out
  /// synchronously here; WatchConnectivity deletes the original as soon as this
  /// method returns.
  func session(_ session: WCSession, didReceive file: WCSessionFile) {
    guard file.metadata?["type"] as? String == "talkAudio",
          let clientId = file.metadata?["clientId"] as? String else { return }

    let copy = FileManager.default.temporaryDirectory
      .appendingPathComponent("talk-\(clientId).m4a")
    try? FileManager.default.removeItem(at: copy)
    do {
      try FileManager.default.copyItem(at: file.fileURL, to: copy)
    } catch {
      replyTranscript(clientId: clientId, text: nil, error: "Audio lost in transfer", ms: 0, onDevice: false)
      return
    }

    // Tell the watch we have it, so its "Sending audio…" can become
    // "Transcribing…" rather than sitting on a stale label.
    _ = push(message: ["type": "audioReceived", "clientId": clientId])
    transcribe(copy, clientId: clientId)
  }
}

// ── speech recognition (this side only) ───────────────────────────────────────

extension WatchBridge {
  private static let recognizer = SFSpeechRecognizer(locale: Locale(identifier: "en-US"))

  /// A clip that arrived in a message payload rather than as a file.
  func receiveInlineAudio(_ audio: Data, clientId: String) {
    let url = FileManager.default.temporaryDirectory
      .appendingPathComponent("talk-\(clientId).m4a")
    do {
      try audio.write(to: url)
    } catch {
      replyTranscript(clientId: clientId, text: nil, error: "Audio lost in transfer", ms: 0, onDevice: false)
      return
    }
    _ = push(message: ["type": "audioReceived", "clientId": clientId])
    transcribe(url, clientId: clientId)
  }

  /// Ask once, at launch, rather than per clip. `requestAuthorization` is cheap after
  /// the first grant but not free, and it sat in the path between the operator
  /// releasing the button and seeing words.
  func primeSpeechAuthorization() {
    guard SFSpeechRecognizer.authorizationStatus() == .notDetermined else { return }
    SFSpeechRecognizer.requestAuthorization { _ in }
  }

  private func transcribe(_ url: URL, clientId: String) {
    let startedAt = Date()
    let finish: (String?, String?, Bool) -> Void = { [weak self] text, error, onDevice in
      try? FileManager.default.removeItem(at: url) // audio is never kept
      let ms = Int(Date().timeIntervalSince(startedAt) * 1000)
      NSLog("[watch] transcribe \(clientId) took \(ms)ms onDevice=\(onDevice) error=\(error ?? "-")")
      self?.replyTranscript(clientId: clientId, text: text, error: error, ms: ms, onDevice: onDevice)
    }

    guard SFSpeechRecognizer.authorizationStatus() == .authorized else {
      // Only ask here if it was never asked — the prompt cannot appear while
      // backgrounded, so failing fast beats hanging on a sheet nobody can see.
      SFSpeechRecognizer.requestAuthorization { status in
        DispatchQueue.main.async {
          if status == .authorized { self.transcribe(url, clientId: clientId) }
          else { finish(nil, "Speech recognition not allowed", false) }
        }
      }
      return
    }
    guard let recognizer = Self.recognizer, recognizer.isAvailable else {
      finish(nil, "Recognizer unavailable", false)
      return
    }

    // Keep the audio on this device when the model is present. It is a ham
    // operator's traffic; there is no reason to hand it to Apple if the phone can do
    // the work itself — and a network round trip is the slower of the two.
    let onDevice = recognizer.supportsOnDeviceRecognition
    let request = SFSpeechURLRecognitionRequest(url: url)
    request.requiresOnDeviceRecognition = onDevice
    request.taskHint = .dictation
    request.shouldReportPartialResults = false // we only ever use the final text

    recognizer.recognitionTask(with: request) { result, error in
      guard let result else {
        if error != nil { finish(nil, "Didn't catch that", onDevice) }
        return
      }
      guard result.isFinal else { return }
      finish(result.bestTranscription.formattedString, nil, onDevice)
    }
  }

  private func replyTranscript(clientId: String, text: String?, error: String?,
                               ms: Int, onDevice: Bool) {
    var payload: [String: Any] = ["type": "transcript", "clientId": clientId,
                                  "ms": ms, "onDevice": onDevice]
    if let text { payload["text"] = text }
    if let error { payload["error"] = error }
    _ = push(message: payload)
  }
}

// ── FlutterStreamHandler ──────────────────────────────────────────────────────

extension WatchBridge: FlutterStreamHandler {
  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    eventSink = events
    flushPendingToDart()
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }
}

File: app/android/app/src/main/kotlin/org/w6sg/aprsmap/watch/WatchBridge.kt
/// Phone side of the Wear OS bridge.
///
/// Owns the Data Layer clients, owns the two Flutter channels, and buffers across the gap
/// between them. Counterpart of `app/ios/Runner/WatchBridge.swift`, talking to the same Dart
/// file (`app/lib/watch_bridge.dart`) over the same channel names and the same payloads — so
/// that one Dart implementation drives both watches and neither can drift.
///
/// **Where this differs from the iOS bridge, and why.**
///
/// 1. *No reply handlers.* WatchConnectivity's `sendMessage` carries a reply channel with a
///    short, unforgiving window, which is why the Swift file has a `replyWatchdog` and a map
///    of held-open handlers. `MessageClient` has none: a message is delivered or it is not.
///    Every answer therefore travels the same way every other push does, as a `sendResult`
///    addressed by `clientId`. That is the path the Swift side already falls back to, so the
///    watch needs no special case — it only ever loses the fast case.
///
/// 2. *No speech recognition.* The Apple Watch has no recogniser and ships audio here to be
///    transcribed. Wear OS watches recognise speech themselves, so there is no `talkAudio`
///    path, no `transcript` reply, and no microphone permission on this side. See
///    `app/android/wear/src/main/kotlin/org/w6sg/aprsmap/wear/SpeechCapture.kt`.
///
/// 3. *A cold process does not start Dart.* WatchConnectivity relaunches the whole iOS app
///    in the background to deliver a message, so `pendingToDart` there is an in-memory queue
///    across a gap of a second or two. Here `WearListenerService` starts a bare process with
///    no Flutter engine in it, and nothing will create one. So the queue is on disk, and it
///    is replayed when Dart next calls `ready` — see README.md ("Wear OS Companion") for
///    what that means for a reply spoken while the phone app is not running.
package org.w6sg.aprsmap.watch

import android.content.Context
import android.net.Uri
import android.util.Log
import com.google.android.gms.tasks.Tasks
import com.google.android.gms.wearable.CapabilityClient
import com.google.android.gms.wearable.CapabilityInfo
import com.google.android.gms.wearable.PutDataMapRequest
import com.google.android.gms.wearable.PutDataRequest
import com.google.android.gms.wearable.Wearable
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.Executors

object WatchBridge : EventChannel.StreamHandler {
    private const val TAG = "watch"

    private const val METHOD_CHANNEL = "org.marsaprs/watch"
    private const val EVENT_CHANNEL = "org.marsaprs/watch/events"

    /// The watch advertises this. Absent means the watch app is not installed.
    const val WEAR_CAPABILITY = "aprs_map_wear"

    /// Phone → watch.
    private const val PATH_CONTEXT = "/aprs/context"
    private const val PATH_LIVE = "/aprs/live"
    private const val PATH_MSG_PREFIX = "/aprs/msg/"

    /// Watch → phone.
    const val PATH_TX = "/aprs/tx"
    const val PATH_TX_QUEUE_PREFIX = "/aprs/txq/"

    const val KEY_JSON = "json"

    /// Durable phone → watch items kept alive at once. The watch deletes each on ingest, so
    /// this only bounds what a watch that has been away for hours comes back to; the same
    /// twenty the context's `recent` list carries.
    private const val DURABLE_CAP = 20

    private const val PREFS = "aprs_watch_bridge"
    private const val KEY_PENDING = "pendingToDart"

    /// Events that arrived before Dart called `ready`. On disk, not in memory: the process
    /// that received them may never run Flutter at all.
    private const val PENDING_CAP = 64

    private lateinit var appContext: Context
    private var methodChannel: MethodChannel? = null
    private var eventSink: EventChannel.EventSink? = null

    /// Last context we pushed, re-sent when the watch becomes reachable again.
    private var lastContext: JSONObject? = null

    /// Durable item paths we have written, oldest first, so old ones can be pruned.
    private val durablePaths = ArrayDeque<String>()

    private val io = Executors.newSingleThreadExecutor()

    private var paired = false
    private var appInstalled = false
    private var reachable = false
    private var watchNodeId: String? = null

    // ── lifecycle ───────────────────────────────────────────────────────────────

    private var listening = false

    /// Called from `MainActivity.onCreate`, and from `WearListenerService` on every callback.
    ///
    /// Both, because either can be the first thing alive in this process: the Activity on a
    /// normal launch, the service when the Data Layer starts us cold to hand over a message.
    /// Idempotent, so the second caller costs nothing.
    @Synchronized
    fun init(context: Context) {
        if (!::appContext.isInitialized) appContext = context.applicationContext
        if (listening) return
        listening = true
        Wearable.getCapabilityClient(appContext)
            .addListener(::onCapabilityChanged, WEAR_CAPABILITY)
        refreshState()
    }

    /// Called once the Flutter engine exists. Safe to call again after an engine restart;
    /// the channels are simply rebuilt.
    fun attach(messenger: BinaryMessenger) {
        MethodChannel(messenger, METHOD_CHANNEL).also { channel ->
            channel.setMethodCallHandler(::handle)
            methodChannel = channel
        }
        EventChannel(messenger, EVENT_CHANNEL).setStreamHandler(this)
    }

    // ── Dart → here ─────────────────────────────────────────────────────────────

    private fun handle(call: MethodCall, result: MethodChannel.Result) {
        val args = (call.arguments as? Map<*, *>)?.let(::toJson) ?: JSONObject()
        when (call.method) {
            "ready" -> {
                flushPendingToDart()
                refreshState()
                result.success(stateMap())
            }

            "status" -> result.success(stateMap())

            "setContext" -> {
                lastContext = args
                result.success(mapOf("ok" to pushContext(args)))
            }

            "pushMessage" -> result.success(mapOf("transport" to push(args)))

            "pushMessages" -> {
                val messages = args.optJSONArray("messages") ?: JSONArray()
                // One envelope, not N: each durable item is a separate wake of the watch
                // app, and a catch-up burst would wake it many times over.
                val transport = if (messages.length() == 0) {
                    "none"
                } else {
                    push(JSONObject().put("batch", messages))
                }
                result.success(mapOf("transport" to transport))
            }

            "sendResult" -> {
                push(args.put("type", "sendResult"))
                result.success(mapOf("ok" to true))
            }

            else -> result.notImplemented()
        }
    }

    private fun stateMap(): Map<String, Any> = mapOf(
        "supported" to true,
        "paired" to paired,
        "appInstalled" to appInstalled,
        "reachable" to reachable,
        "activated" to true,
    )

    // ── here → watch ────────────────────────────────────────────────────────────

    /// Latest-value-wins state: token, destination, conversation list, speak flag. Survives
    /// the watch app not running, and coalesces if we push faster than the link drains.
    /// Returns false when there is nothing to talk to.
    private fun pushContext(dict: JSONObject): Boolean {
        if (!::appContext.isInitialized) return false
        // One fixed path, so a second push replaces the first rather than queueing behind
        // it. That is the whole of `updateApplicationContext` expressed in the Data Layer.
        putDurable(PATH_CONTEXT, dict, prune = false)
        Log.d(TAG, "context pushed (sharing=${dict.opt("sharing")} lastId=${dict.opt("lastId")})")
        return true
    }

    /// A message (or a `batch` of them, or a receipt, or a send result).
    ///
    /// `MessageClient` is the low-latency path when the watch app is up; a `DataClient` item
    /// is the durable one — it survives both processes dying and starts the watch's listener
    /// service in the background to receive. The watch dedupes by message id, so the
    /// fallback racing the fast path is harmless.
    /// Returns the transport it will use, not the transport that succeeded. The Data Layer
    /// calls all block, and this is invoked from the Flutter method-channel thread, which is
    /// the main thread — so the work happens on `io` and the answer to Dart is a prediction.
    /// Dart only logs it; the watch is the thing that reports what actually arrived.
    private fun push(payload: JSONObject): String {
        if (!::appContext.isInitialized) return "none"
        val node = watchNodeId
        // A receipt is about something happening right now. Delivered twenty minutes late,
        // after the wrist comes back into range, "message delivered" is not news — it is a
        // spoken interruption about a message the operator has long since moved on from. The
        // fast path or nothing.
        val receipt = payload.optString("type") == "receipt"
        if (node == null && receipt) {
            Log.d(TAG, "receipt dropped: watch not reachable")
            return "none"
        }

        if (node != null) {
            val bytes = payload.toString().toByteArray(Charsets.UTF_8)
            io.execute {
                val ok = runCatching {
                    Tasks.await(
                        Wearable.getMessageClient(appContext).sendMessage(node, PATH_LIVE, bytes)
                    )
                }.isSuccess
                if (ok || receipt) return@execute
                Log.d(TAG, "live push failed, falling back to a durable item")
                putDurable(PATH_MSG_PREFIX + durableSuffix(payload), payload, prune = true)
            }
            return "message"
        }

        putDurable(PATH_MSG_PREFIX + durableSuffix(payload), payload, prune = true)
        return "dataItem"
    }

    /// A stable, unique path per payload. Stable so that the same message pushed twice is
    /// one item rather than two; unique so that two different messages never collide, which
    /// would silently drop the second — the Data Layer only notifies on change.
    private fun durableSuffix(payload: JSONObject): String = when {
        payload.has("id") -> payload.optInt("id").toString()
        payload.has("clientId") -> "s" + payload.optString("clientId")
        payload.has("batch") -> {
            val batch = payload.optJSONArray("batch") ?: JSONArray()
            var max = 0
            for (i in 0 until batch.length()) {
                max = maxOf(max, batch.optJSONObject(i)?.optInt("id") ?: 0)
            }
            "b$max"
        }
        else -> "x" + System.currentTimeMillis()
    }

    /// A host-less `wear://` URI, which the Data Layer reads as "this path on every node".
    /// Building it by hand risks `wear:///aprs/...` versus `wear://aprs/...`, and the second
    /// silently matches nothing.
    private fun wearUri(path: String): Uri = Uri.Builder()
        .scheme(PutDataRequest.WEAR_URI_SCHEME)
        .path(path)
        .build()

    private fun putDurable(path: String, payload: JSONObject, prune: Boolean) {
        io.execute {
            runCatching {
                val request = PutDataMapRequest.create(path)
                request.dataMap.putString(KEY_JSON, payload.toString())
                Tasks.await(
                    Wearable.getDataClient(appContext)
                        .putDataItem(request.asPutDataRequest().setUrgent())
                )
            }.onFailure { Log.w(TAG, "putDataItem $path failed: ${it.message}") }

            if (!prune) return@execute
            synchronized(durablePaths) {
                durablePaths.remove(path)
                durablePaths.addLast(path)
                while (durablePaths.size > DURABLE_CAP) {
                    val old = durablePaths.removeFirst()
                    runCatching {
                        Tasks.await(
                            Wearable.getDataClient(appContext).deleteDataItems(wearUri(old))
                        )
                    }
                }
            }
        }
    }

    // ── watch → Dart ────────────────────────────────────────────────────────────

    /// Route an inbound watch payload to Dart, or to disk if Dart is not there.
    ///
    /// Called from `WearListenerService`, which may be the only thing alive in this process.
    fun receive(payload: JSONObject) {
        val event = if (payload.has("type")) payload else JSONObject(payload.toString()).put("type", "unknown")
        val sink = eventSink
        if (sink != null) {
            android.os.Handler(android.os.Looper.getMainLooper()).post {
                runCatching { sink.success(toMap(event)) }
            }
            return
        }

        // Dart is not listening. Park it, and — for a send — tell the watch so, rather than
        // leaving its Outbox row claiming to be in flight forever.
        parkForDart(event)
        val clientId = event.optString("clientId")
        if (event.optString("type") == "send" && clientId.isNotEmpty()) {
            push(
                JSONObject()
                    .put("type", "sendResult")
                    .put("clientId", clientId)
                    .put("queued", true)
            )
        }
    }

    private fun parkForDart(event: JSONObject) {
        if (!::appContext.isInitialized) return
        val prefs = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val existing = runCatching { JSONArray(prefs.getString(KEY_PENDING, "[]")) }
            .getOrDefault(JSONArray())
        val out = JSONArray()
        // Cap the buffer: a watch that keeps talking to an app that never runs must not grow
        // this without bound.
        val start = maxOf(0, existing.length() - (PENDING_CAP - 1))
        for (i in start until existing.length()) out.put(existing.get(i))
        out.put(event)
        prefs.edit().putString(KEY_PENDING, out.toString()).apply()
    }

    private fun flushPendingToDart() {
        val sink = eventSink ?: return
        if (!::appContext.isInitialized) return
        val prefs = appContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val queued = runCatching { JSONArray(prefs.getString(KEY_PENDING, "[]")) }
            .getOrDefault(JSONArray())
        if (queued.length() == 0) return
        prefs.edit().remove(KEY_PENDING).apply()
        for (i in 0 until queued.length()) {
            val event = queued.optJSONObject(i) ?: continue
            runCatching { sink.success(toMap(event)) }
        }
    }

    // ── link state ──────────────────────────────────────────────────────────────

    private fun onCapabilityChanged(info: CapabilityInfo) {
        applyCapability(info)
        emitWatchState()
        // The watch stops its own direct polling the moment we are reachable, so it needs
        // current state immediately rather than at the next natural push.
        if (reachable) lastContext?.let { pushContext(it) }
    }

    private fun applyCapability(info: CapabilityInfo) {
        appInstalled = info.nodes.isNotEmpty()
        val nearby = info.nodes.firstOrNull { it.isNearby }
        watchNodeId = nearby?.id
        reachable = nearby != null
    }

    fun refreshState() {
        if (!::appContext.isInitialized) return
        io.execute {
            runCatching {
                paired = Tasks.await(Wearable.getNodeClient(appContext).connectedNodes).isNotEmpty()
                val info = Tasks.await(
                    Wearable.getCapabilityClient(appContext)
                        .getCapability(WEAR_CAPABILITY, CapabilityClient.FILTER_ALL)
                )
                applyCapability(info)
            }.onFailure { Log.w(TAG, "state refresh failed: ${it.message}") }
            emitWatchState()
        }
    }

    private fun emitWatchState() {
        val sink = eventSink ?: return
        val event = stateMap().toMutableMap()
        event["type"] = "watchState"
        android.os.Handler(android.os.Looper.getMainLooper()).post {
            runCatching { sink.success(event) }
        }
    }

    // ── EventChannel.StreamHandler ──────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        flushPendingToDart()
        emitWatchState()
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ── JSON ⇄ Flutter's standard codec ─────────────────────────────────────────

    /// Flutter's codec hands Dart maps over as `Map<*, *>` of boxed primitives. Everything
    /// downstream of here speaks JSON, because the Data Layer, the watch and the server all
    /// do — so the conversion happens once, at the boundary, rather than being repeated at
    /// every field.
    private fun toJson(map: Map<*, *>): JSONObject {
        val out = JSONObject()
        for ((key, value) in map) {
            val name = key as? String ?: continue
            // Dropped rather than written as JSON null. Absent means null by contract on the
            // watch side, and the context is always a complete snapshot, so dropping them
            // loses nothing — while a null that reaches `optInt` comes back as 0, which is a
            // valid conversation id and would aim a reply at nothing.
            if (value == null) continue
            out.put(name, toJsonValue(value))
        }
        return out
    }

    private fun toJsonValue(value: Any?): Any = when (value) {
        null -> JSONObject.NULL
        is Map<*, *> -> toJson(value)
        is List<*> -> JSONArray().also { array ->
            value.forEach { array.put(toJsonValue(it)) }
        }
        else -> value
    }

    private fun toMap(obj: JSONObject): Map<String, Any?> {
        val out = HashMap<String, Any?>()
        for (key in obj.keys()) out[key] = fromJsonValue(obj.get(key))
        return out
    }

    private fun fromJsonValue(value: Any?): Any? = when (value) {
        JSONObject.NULL, null -> null
        is JSONObject -> toMap(value)
        is JSONArray -> (0 until value.length()).map { fromJsonValue(value.get(it)) }
        else -> value
    }
}
